### ISSUE TITLE
Erratic results with MobileNet Cifar10

### ISSUE THREAD

**Issue by piyush-das** [ozora-ogino/asoftmax-tf#2]
Hi, I was experimenting with AsoftMax and mobilenet for Cifar10, but it seems the training results are very erratic and can't be explained when we use the Asoftmax layer. Following are the scripts and results which I have obtained.

```python
import numpy as np
import tensorflow as tf
from keras_arcface import AsoftMax
from tensorflow.keras import regularizers
from tensorflow.keras.applications.mobilenet import MobileNet


class CifarModel(tf.keras.models.Model):
    def __init__(self):
        super(CifarModel, self).__init__()
        # self.layer_1 = tf.keras.layers.Dense(32, activation="relu")
        # self.layer_2 = tf.keras.layers.Dense(10)
        self.basemodel = MobileNet(input_shape=(32,32,3), alpha=0.25, weights="imagenet", include_top=False)
        self.bn = tf.keras.layers.BatchNormalization()
        self.dropout = tf.keras.layers.Dropout(rate=0.5)
        self.flatten = tf.keras.layers.Flatten()
        self.dense = tf.keras.layers.Dense(64)
        self.bn2 = tf.keras.layers.BatchNormalization()

        self.out = AsoftMax(
            n_classes=10,
            scale=30.,
            regularizer=regularizers.l2(1e-4),
        )
        # self.out = tf.keras.layers.Dense(10, activation="softmax")

    def call(self, x, training=False):
        if training:
            x, y = x[0], x[1]
        x = self.basemodel(x)
        x = self.bn(x)
        x = self.dropout(x)
        x = self.flatten(x)
        x = self.dense(x)
        x = self.bn2(x)

        if training:
            # When training, you need to pass label to AsoftMax
            out = self.out([x, y])
        else:
            out = self.out(x)
        return out


def preprocessing(data, labels):
    data = data.astype('float32')
    data /= 127.5
    data -= 1.
    labels = tf.keras.utils.to_categorical(labels, 10)
    return data, labels


def acc(pred, label):
    return len(pred[label == np.argmax(pred, axis=1)]) / len(pred)


if __name__ == "__main__":
    epochs = 10
    batch_size = 64

    # (x_train, y_train), (x_test, y_test) = tf.keras.datasets.mnist.load_data()
    (x_train, y_train), (x_test, y_test) = tf.keras.datasets.cifar10.load_data()
    x_train, y_training = preprocessing(x_train, y_train)
    x_test, y_testing = preprocessing(x_test, y_test)

    model = CifarModel()
    optimizer = tf.keras.optimizers.Adam()
    # Please note that you need to use categorical_crossentropy. (NOT sparse_categorical_crossentropy)
    loss = tf.keras.losses.categorical_crossentropy

    model.compile(loss=loss, optimizer=optimizer, metrics=["acc"])

    model.fit(
        [x_train, y_training],
        y_training,
        epochs=5,
        validation_data=(x_test, y_testing),
        batch_size=batch_size,
    )


    pred = model.predict(x_test)
    accuracy = acc(pred, y_testing)
    print(f"Accuracy: {accuracy*100}%")  # Accuracy: 96.07%
```

Here is the output for the same
```log
Epoch 1/5
782/782 [==============================] - 10s 9ms/step - loss: 15.2321 - acc: 0.0494 - val_loss: 2.2742 - val_acc: 0.2670
Epoch 2/5
782/782 [==============================] - 6s 8ms/step - loss: 1.9333 - acc: 0.7268 - val_loss: 2.3043 - val_acc: 0.0961
Epoch 3/5
782/782 [==============================] - 6s 8ms/step - loss: 0.5324 - acc: 0.9573 - val_loss: 2.3043 - val_acc: 0.0966
Epoch 4/5
782/782 [==============================] - 6s 8ms/step - loss: 0.4379 - acc: 0.9686 - val_loss: 2.3043 - val_acc: 0.1011
Epoch 5/5
782/782 [==============================] - 6s 8ms/step - loss: 0.4340 - acc: 0.9676 - val_loss: 2.3043 - val_acc: 0.1042
```

As can be observed the training accuracy jumps exponentially (abnormally high) and the validation gets stuck.  While if I replace the asoftmax layer with a Dense activation, the training proceeds as follows :
```log
Epoch 1/5
782/782 [==============================] - 9s 8ms/step - loss: 1.7380 - acc: 0.3929 - val_loss: 1.1625 - val_acc: 0.5840
Epoch 2/5
782/782 [==============================] - 6s 8ms/step - loss: 1.1664 - acc: 0.5863 - val_loss: 1.0570 - val_acc: 0.6228
Epoch 3/5
782/782 [==============================] - 6s 8ms/step - loss: 1.0368 - acc: 0.6356 - val_loss: 0.9599 - val_acc: 0.6562
Epoch 4/5
782/782 [==============================] - 6s 8ms/step - loss: 0.9558 - acc: 0.6659 - val_loss: 0.9735 - val_acc: 0.6564
Epoch 5/5
782/782 [==============================] - 6s 8ms/step - loss: 0.8942 - acc: 0.6882 - val_loss: 0.9394 - val_acc: 0.6704
```

Could you please tell me where I am going wrong. Thanks

**Comment by ozora-ogino** [comment]
@piyush-das  Thank you for sharing the issue.
I have no idea what is wrong...
Let me check how to fix it.
If you can debug it, it's really helpful.

**Comment by dimka11** [comment]
I can confirm, that this realization ArcFace loss doesn't works. Train accuracy is 1. Loss is not drop. Validation accuracy is not improving.
